lib/promoted/ruby/client/pager.rb

```diff
@@ -13,34 +13,43 @@ def initialize(message, default_insertions_page)
         class Pager
             def validate_paging(insertions, retrieval_insertion_offset, paging)
               if paging && paging[:offset]
-                offset = [0, paging[:offset]].max
-                if offset >= insertions.length
-                  raise InvalidPagingError.new("Invalid page offset (insertion size #{insertions.length}, offset #{offset})", [])
-                end
-                if offset < retrieval_insertion_offset
-                  raise InvalidPagingError.new("Invalid page offset (retrieval_insertion_offset #{retrieval_insertion_offset}, offset #{offset})", [])
-                end
+                offset, retrieval_insertion_offset, index = _sanitize_offsets(retrieval_insertion_offset, paging)
+                _validate_paging(insertions, retrieval_insertion_offset, offset, index)
               end
             end
 
-            def apply_paging(insertions, retrieval_insertion_offset, paging = nil)
-              begin
-                validate_paging(insertions, retrieval_insertion_offset, paging)
-              rescue InvalidPagingError => err
-                # This is invalid input, stop it before it goes to the server.
-                return err.default_insertions_page
+            def _sanitize_offsets(retrieval_insertion_offset, paging)
+              offset = [0, paging[:offset]].max
+              retrieval_insertion_offset = [0, retrieval_insertion_offset].max
+              index = [0, offset - retrieval_insertion_offset].max
+              return [offset, retrieval_insertion_offset, index]
+            end
+
+            def _validate_paging(insertions, retrieval_insertion_offset, offset, index)
+              if offset < retrieval_insertion_offset
+                raise InvalidPagingError.new("Invalid page offset (retrieval_insertion_offset #{retrieval_insertion_offset}, offset #{offset})", [])
+              end
+              if index >= insertions.length
```

Comment:
We need to use index instead here (since index is relative to the insertions list).

Reply:
Yeah, when I was writing this logic for the .NET SDK I drafted a lot of comments about what each index precisely was,  but it was really hard to explain

Reply:
Longer-term, I'd guess we'll rip out some of this validation logic.  It moves a corner case that we could handle in the SDK to a problem for the caller.